File: cons.c
#include "cons.h"
#include "unsafe.h"

#include <assert.h>
#include <stdlib.h>

typedef struct {
  const Cons *left;
  const Cons *right;
} Tuple;

struct _Cons {
  enum { ETuple, ELeaf } type;
  union {
    Tuple tuple;
    void *leaf;
  };
};

Cons *cons_new_tuple(const Cons *left, const Cons *right) {
  Cons *c = malloc(sizeof(Cons));
  c->type = ETuple;
  c->tuple.left = left;
  c->tuple.right = right;
  return c;
}

Cons *cons_new_leaf(void *leaf) {
  Cons *c = malloc(sizeof(Cons));
  c->type = ELeaf;
  c->leaf = leaf;
  return c;
}

void *cons_get_value(const Cons *self) {
  return cons_is_leaf(self) ? self->leaf : NULL;
}

const Cons *cons_get_left(const Cons *self) {
  return cons_is_tuple(self) ? self->tuple.left : NULL;
}

const Cons *cons_get_right(const Cons *self) {
  return cons_is_tuple(self) ? self->tuple.right : NULL;
}

bool cons_is_leaf(const Cons *self) { return self->type == ELeaf; }

bool cons_is_tuple(const Cons *self) { return self->type == ETuple; }

void cons_post_order(const Cons *self, Closure *c) {
  switch (self->type) {
  case ELeaf:
    closure_apply(c, void_cast(self));
    break;
  case ETuple:
    cons_post_order(cons_get_left(self), c);
    cons_post_order(cons_get_right(self), c);
    closure_apply(c, void_cast(self));
    break;
  default:
    assert(false);
  }
}

static void *_cons_free(Closure *self, void *args) {
  free(args);
  return NULL;
}

void cons_free(Cons *self) {
  Closure free_closure = {.apply = _cons_free};
  cons_post_order(self, &free_closure);
}

File: Cons.java
public interface Cons<T> {
	public static class Tuple<T> implements Cons<T> {
		private final Cons<T> left;
		private final Cons<T> right;
		public Cons<T> left() { return left; }
		public Cons<T> right() { return right; }
		private Tuple(Cons<T> left, Cons<T> right) { this.left = left; this.right = right; }
	}
	public static class Leaf<T> implements Cons<T> {
		private final T value;
		private Leaf(T value) { this.value = value; }
		public T value() { return this.value; }
	}

	public static <T> Tuple<T> newTuple(Cons<T> left, Cons<T> right) { return new Tuple<>(left, right); }
	public static <T> Leaf<T> newLeaf(T value) { return new Leaf<>(value); }

	default boolean isLeaf() { return this instanceof Leaf; }
	default boolean isTuple() { return ! isLeaf(); }
	default Tuple<T> asTuple() { assert isTuple(); return (Tuple<T>)this; }
	default Leaf<T> asLeaf() { assert isLeaf(); return (Leaf<T>)this; }
}
